File: module_4/python/sort_list.py
"""
Title: Sort list

Problem: Sort a linked list in O(n log n) time using constant space complexity.


Execution: python sort_list.py
"""
import unittest


class ListNode:
    """Definition for singly-linked list."""

    def __init__(self, x: int):
        """Set value and next for ListNode."""
        self.val = x
        self.next = None


def merge(h1: ListNode, h2: ListNode) -> ListNode:
    """Merge two ListNode objects."""
    dummy = tail = ListNode(None)
    while h1 and h2:
        if h1.val < h2.val:
            tail.next, tail, h1 = h1, h1, h1.next
        else:
            tail.next, tail, h2 = h2, h2, h2.next

    tail.next = h1 or h2
    return dummy.next


def sort_list(head: ListNode) -> ListNode:
    """Sort ListNode objects."""
    if not head or not head.next:
        return head

    pre, slow, fast = None, head, head
    while fast and fast.next:
        pre, slow, fast = slow, slow.next, fast.next.next
    pre.next = None

    return merge(*map(sort_list, (head, slow)))


def compare_lists(l1: ListNode, l2: ListNode) -> bool:
    """Compare equality between two ListNode objects."""
    while l1:
        if l1.val != l2.val:
            return False
        l1 = l1.next
        l2 = l2.next
    return True


class TestSortList(unittest.TestCase):
    """Unit test for sort_list."""

    def test_1(self):
        ll = ListNode(4)
        ll.next = ListNode(2)
        ll.next.next = ListNode(1)
        ll.next.next.next = ListNode(3)

        out_ll = ListNode(1)
        out_ll.next = ListNode(2)
        out_ll.next.next = ListNode(3)
        out_ll.next.next.next = ListNode(4)

        self.assertEqual(compare_lists(sort_list(ll), out_ll), True)


if __name__ == '__main__':
    unittest.main()

File: module_4/java/SortList.java
/*
 *   Title: Sort List
 *
 *   Problem:Sort a linked list in O(n log n) time using constant space
 *   complexity.
 *
 *   Execution: javac SortList.java && java SortList
 */
import java.util.*;


public class SortList {

    public static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) { val = x; }
    }

    public static ListNode sortList(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        int n = 0;
        while (head != null) {
            head = head.next;
            n++;
        }
        
        for (int step = 1; step < n; step <<= 1) {
            ListNode prev = dummy;
            ListNode cur = dummy.next;
            while (cur != null) {
                ListNode left = cur;
                ListNode right = split(left, step);
                cur = split(right, step);
                prev = merge(left, right, prev);
            } 
        }
        
        return dummy.next;
    }
    
    private static ListNode split(ListNode head, int step) {
        if (head == null) return null;
    	
        for (int i = 1; head.next != null && i < step; i++) {
            head = head.next;
        }
        
        ListNode right = head.next;
        head.next = null;
        return right;
    }
    
    private static ListNode merge(ListNode left, ListNode right, ListNode prev) {
        ListNode cur = prev;
        while (left != null && right != null) {
            if (left.val < right.val) {
                cur.next = left;
                left = left.next;
            }
            else {
                cur.next = right;
                right = right.next;
            }
            cur = cur.next;
        }
        
        if (left != null) cur.next = left;
        else if (right != null) cur.next = right;
        while (cur.next != null) cur = cur.next;
        return cur;
    }

    public static boolean compareLists(ListNode l1, ListNode l2) {
        while (l1 != null) {
            if (l1.val != l2.val) {
                return false;
            }
            l1 = l1.next;
            l2 = l2.next;
        }
        return true;
    }

    public static void main(String[] args) {
        ListNode l1 = new ListNode(4);
        l1.next = new ListNode(2);
        l1.next.next = new ListNode(1);
        l1.next.next.next = new ListNode(3);

        ListNode expected_l1 = new ListNode(1);
        expected_l1.next = new ListNode(2);
        expected_l1.next = new ListNode(3);
        expected_l1.next = new ListNode(4);

        assert compareLists(sortList(l1), expected_l1) == true;

        ListNode l2 = new ListNode(1);
        l2.next = new ListNode(5);
        l2.next.next = new ListNode(3);
        l2.next.next.next = new ListNode(4);
        l2.next.next.next.next = new ListNode(0);

        ListNode expected_l2 = new ListNode(1);
        expected_l2.next = new ListNode(0); 
        expected_l2.next.next = new ListNode(3);
        expected_l2.next.next.next = new ListNode(4);
        expected_l2.next.next.next.next = new ListNode(5);

        assert compareLists(sortList(l2), expected_l2) == true;

        System.out.println("Passed all test cases");
    }
    
}
